### PR TITLE
Il-to-IL pass: Animate

### PR DESCRIPTION
**Pull request by f52985**
Il-to-IL pass: Animate

This transformation will

1. make the equality conditions explicit as assignment conditions, if they are indeed assignments, and
2. reorder the premises that resembles the actual order of execution in interpreter

ex) `--if x = [c]` -> `--where [c] := x`

This PR adds a new constructor of`Il.Ast.AssignPr` to the type `Il.Ast.premise'`

### CODE REVIEW

File: spectec/src/il/ast.ml [hunk 1]
```diff
@@ -156,6 +156,7 @@ and premise = premise' phrase
 and premise' =
   | RulePr of id * mixop * exp                        (* premise *)
   | IfPr of exp                                       (* side condition *)
+  | AssignPr of exp * exp                             (* assignment *)
```
Comment (rossberg) [inline spectec/src/il/ast.ml:159]: Nit: can we call this LetPr?

File: spectec/src/il/print.ml [hunk 2]
```diff
@@ -198,6 +198,7 @@ let rec string_of_prem prem =
   match prem.it with
   | RulePr (id, op, e) -> id.it ^ ": " ^ string_of_exp {e with it = MixE (op, e)}
   | IfPr e -> "if " ^ string_of_exp e
+  | AssignPr (e1, e2) -> "where " ^ string_of_exp e1 ^ " := " ^ string_of_exp e2
```
Comment (rossberg) [inline spectec/src/il/print.ml:201]:
```suggestion
  | AssignPr (e1, e2) -> "where " ^ string_of_exp e1 ^ " = " ^ string_of_exp e2
```

File: spectec/src/il/validation.ml [hunk 3]
```diff
@@ -498,6 +499,8 @@ let rec valid_prem env prem =
     valid_expmix env mixop e (find "relation" env.rels id) e.at
   | IfPr e ->
     valid_exp env e (BoolT $ e.at)
+  | AssignPr (e1, e2) ->
+    valid_exp env (CmpE (EqOp, e1, e2) $$ no_region % (BoolT $ no_region))  (BoolT $ prem.at)
```
Comment (rossberg) [inline spectec/src/il/validation.ml:503]:
```suggestion
    valid_exp env (CmpE (EqOp, e1, e2) $$ prem.at % (BoolT $ prem.at)) (BoolT $ prem.at)
```

**Review by rossberg — COMMENTED** [review]
Mostly nits for IL, leaving middlend for others to review.

File: spectec/src/il/ast.ml (hunk 1 above)
Comment (f52985) [reply spectec/src/il/ast.ml:159]: Yes, that sounds better.

**Pull request merged**